src/btreemap/iter.rs

```diff
@@ -260,6 +348,87 @@ where
             }
         }
     }
+
+    // Iterates to find the next back element in the requested range.
+    // If it exists, `map` is applied to that element and the result is returned.
+    fn next_back_map<T, F: Fn(&Node<K>, usize) -> T>(&mut self, map: F) -> Option<T> {
+        if !self.backward_cursors_initialized {
+            self.initialize_backward_cursors();
+        }
+
+        // If the cursors are empty. Iteration is complete.
+        match self.backward_cursors.pop()? {
+            Cursor::Address(address) => {
+                if address != NULL {
+                    // Load the node at the given address, and add it to the cursors.
+                    let node = self.map.load_node(address);
+                    if let Some(next) = match node.node_type() {
+                        // Iterate on internal nodes starting from the first child.
+                        NodeType::Internal if node.children_len() > 0 => {
+                            Some(Index::Child(node.children_len() - 1))
+                        }
+                        // Iterate on leaf nodes starting from the first entry.
+                        NodeType::Leaf if node.entries_len() > 0 => {
+                            Some(Index::Entry(node.entries_len() - 1))
+                        }
+                        _ => None,
+                    } {
+                        self.backward_cursors.push(Cursor::Node { next, node });
+                    }
+                }
+                self.next_back_map(map)
+            }
+
+            Cursor::Node {
+                node,
+                next: Index::Child(child_idx),
+            } => {
+                let child_address = node.child(child_idx);
+
+                if 0 < child_idx && child_idx <= node.entries_len() {
+                    // After iterating on the child, iterate on the previous _entry_ in this node.
+                    self.backward_cursors.push(Cursor::Node {
+                        node,
+                        next: Index::Entry(child_idx - 1),
+                    });
+                }
+
+                // Add the child to the top of the cursors to be iterated on first.
+                self.backward_cursors.push(Cursor::Address(child_address));
+
+                self.next_back_map(map)
+            }
+
+            Cursor::Node {
+                node,
+                next: Index::Entry(entry_idx),
+            } => {
```

Comment:
Should we also have a short-circuit here like in the `next_map` where we do something like
```
if entry_idx <= 0 {
    // No more entries to iterate on in this node.
    return self.next_map(map);
}
```

I think this might then simplify the code later below because you could just use `entry_idx - 1` without worrying about it and it would be more similar to the other case (which makes it easier to inspect correctness). Unless there's something else I'm missing of course.

Reply:
In the `next_map` case, the check on if the index exceeds the entries was never done when pushing cursors, it was then handled by the `if entry_idx >= node.entries_len() { .. }` check.
Whereas in the `next_map_back` the check was being done before pushing the cursor to prevent underflow (entry_id of 0 is valid).
I've now modified it so that the checks always happen before pushing the cursors when iterating in either direction.
This means the check happens in 2 places rather than 1 in the forward case, but it also means we skip pushing and popping unnecessary cursors that we know won't contain any items.